File: now/week5/4-7.js
// 给定一个 没有重复 数字的序列，返回其所有可能的全排列。(l - 46)
// 1. 回溯. 在每次访问的时候, 判断数组是否访问过, 访问过的数字跳过.

var permute = function(nums) {
  if (nums.length === 0) return [];
  const ans = [];
  const visited = Array(nums.length).fill(false);
  const res = [];
  trackback(res);
  return ans;
  function trackback(temp) {
    if (temp.length === nums.length) {
      const currentResult = JSON.parse(JSON.stringify(temp));
      ans.push(currentResult);
      return;
    }

    for (let i = 0; i < nums.length; i++) {
      if (!visited[i]) {
        visited[i] = true;
        temp.push(nums[i]);
        trackback(temp);
        temp.pop();
        visited[i] = false;
      }
    }
  }
};

File: now/week5/47.java
class Solution{
  List<List<Integer>> res=new ArrayList<>();
  public List<List<Integer>> permuteUnique(int[] nums) {
      Arrays.sort(nums);
      List<Integer> temp=new ArrayList<>();
      boolean[] visited=new boolean[nums.length];
      for (int i=0;i<nums.length;i++)
          visited[i]=false;
      helper(nums,visited,temp);
      return res;

  }
  public void helper(int[] nums, boolean[] visited,List<Integer> temp)
  {
      if(temp.size()==nums.length)
      {
          res.add(new ArrayList<>(temp));
          return;
      }
      for(int i=0;i<nums.length;i++)
      {
          if(!visited[i])
          {
              visited[i]=true;
              temp.add(nums[i]);
              helper(nums,visited,temp);
              temp.remove(temp.size()-1);
              visited[i]=false;
              while(i+1<nums.length && nums[i]==nums[i+1])
                  i++;
          }

      }
  }
}
